### DISCUSSION TITLE
New features :)

### DISCUSSION THREAD

**Discussion by Romanitho** [Romanitho/Winget-Install-GUI#3] [Show and tell]
Hey @KnifMelti

New features implemented in the GUI tool :)

Maybe it can be useful for list creation when WAU is installed silently

**Comment by KnifMelti** [comment]
Nice with list creation!
But, we noticed so much security related things in many installs/uninstalls that we test them extensively and then whitelist them (hoping upgrades doesn't suddenly inject security issues for the whitelisted ones, because of the system context).
So, to just make a list and install it is not an option for us!
But it's great thinking 😀

**Reply by KnifMelti** [reply]
You could in the GUI also implement load from list, then sharing a bunch of great applications to a friend becomes easy.

**Reply by Romanitho** [reply]
You can use it to create white list for WAU :)

Load function, not bad idea.

**Reply by KnifMelti** [reply]
Yes I read that now 😁
But we create and delete on the fly

**Reply by Romanitho** [reply]
It can be useful for standalone usage, or home PC :)

**Reply by KnifMelti** [reply]
Very true!

**Reply by KnifMelti** [reply]
Winget has an export/import feature too

**Reply by Romanitho** [reply]
so, no GUI necessary :p

**Reply by KnifMelti** [reply]
I meant making it possible to import a native exported list in your GUI 👍 and maybe export to a native list for importing in winget...

**Comment by KnifMelti** [comment]
@Romanitho 
Wow!
Now it became extremely useful 💯 (but maybe not in your indented way,,,)
As a virtual developer/tester I've implemented the solution from @damienvanrobaeys Run-In-Sandbox (https://github.com/microsoft/Windows-Sandbox-Utilities) with a custom Task Schedule that disables the virtual switch in Windows Sandbox making it secure enough to let some of our adult users use it to test/classifie applications in.
In the future a GPO for disallowing the network interface is coming from Microsoft and then my Task Schedule will be deleted.

Anyway (I don't have the switch blocking Task in my setups):
Now, with a simple click in WiGui.exe (Install WAU) after the initial failure (Winget doesn't exist in Windows Sandbox) the Winget parts/dependencies gets installed and after that I can run WiGui in System Context via Nir Sofers AdvancedRun to install anything! (https://www.nirsoft.net/utils/advanced_run.html).

Proposals:
- get rid of the errors in Windows Sandbox (see below)
- log to WAU logs (same as with Winget-Install)
- make the Search button respond to Enter key in the search field
- make multiple choices in the result list possible (Shift/Ctrl marking)
- don't end WiGui after clicking on the Install button

Great solution!

**Reply by KnifMelti** [reply]
The error dialogs in Windows Sandbox (when starting WiGui.exe if WAU is not installed) are:

In Admin Context:
```
---------------------------
WiGui.exe
---------------------------
Deployment failed with HRESULT: 0x80073CF0, Package could not be opened.



error 0x80070002: Opening the package from location Microsoft.VCLibs.x64.14.00.Desktop.appx failed.

NOTE: For additional information, look for [ActivityId] baca7532-50e3-0003-641a-cbbae350d801 in the Event Log or use the command line Get-AppPackageLog -ActivityID baca7532-50e3-0003-641a-cbbae350d801

---------------------------
OK   
---------------------------


---------------------------
WiGui.exe
---------------------------
Deployment failed with HRESULT: 0x80073CF0, Package could not be opened.



error 0x80070002: Opening the package from location Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle failed.

NOTE: For additional information, look for [ActivityId] baca7532-50e3-0001-e60d-cbbae350d801 in the Event Log or use the command line Get-AppPackageLog -ActivityID baca7532-50e3-0001-e60d-cbbae350d801

---------------------------
OK   
---------------------------
```

In System Context:
```
---------------------------
WiGui.exe
---------------------------
Deployment failed with HRESULT: 0x80073CF9, Install failed. Please contact your software vendor. (Exception from HRESULT: 0x80073CF9)
Deployment Add operation rejected on package Microsoft.UI.Xaml.2.7_7.2109.13004.0_x64__8wekyb3d8bbwe from: Microsoft.UI.Xaml.2.7.appx install request because the Local System account is not allowed to perform this operation.

NOTE: For additional information, look for [ActivityId] baca7532-50e3-0002-231d-cbbae350d801 in the Event Log or use the command line Get-AppPackageLog -ActivityID baca7532-50e3-0002-231d-cbbae350d801

---------------------------
OK   
---------------------------

---------------------------
WiGui.exe
---------------------------
Deployment failed with HRESULT: 0x80073CF0, Package could not be opened.



error 0x80070002: Opening the package from location Microsoft.VCLibs.x64.14.00.Desktop.appx failed.

NOTE: For additional information, look for [ActivityId] baca7532-50e3-0003-412b-cbbae350d801 in the Event Log or use the command line Get-AppPackageLog -ActivityID baca7532-50e3-0003-412b-cbbae350d801

---------------------------
OK   
---------------------------

---------------------------
WiGui.exe
---------------------------
Deployment failed with HRESULT: 0x80073CF0, Package could not be opened.



error 0x80070002: Opening the package from location Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle failed.

NOTE: For additional information, look for [ActivityId] baca7532-50e3-0001-803a-cbbae350d801 in the Event Log or use the command line Get-AppPackageLog -ActivityID baca7532-50e3-0001-803a-cbbae350d801

---------------------------
OK   
---------------------------

```

**Reply by Romanitho** [reply]
Can be easy to fix (I guess), as the prerequisites are not exactly configured identically in WAU and WiGui (same as Winget-AllInOne project).

**Reply by Romanitho** [reply]
Regarding your proposals :  

- get rid of the errors in Windows Sandbox (see below)log to WAU logs (same as with Winget-Install)

> Can try to fix

- make the Search button respond to Enter key in the search field

> I will try to see how i can do that, as it's my first real powershell GUI 😅😅

- make multiple choices in the result list possible (Shift/Ctrl marking)

> I think it's not that easy to implement, and you can submit apps one by one in the droplist. But later ? 😇

- don't end WiGui after clicking on the Install button

> What are you thinking about? For which use case? Maybe a `Close` button instead `Cancel` then.

**Reply by KnifMelti** [reply]
In an sanboxed environment with WiGui open started in system context I'm testing a lot of things. Then I don't need to start AdvancedRun, start WiGui again as system.. ..one step less to take 😊

**Reply by Romanitho** [reply]
Nice :)

**Reply by KnifMelti** [reply]
https://docs.microsoft.com/en-us/powershell/scripting/samples/multiple-selection-list-boxes?view=powershell-5.1

**Reply by Romanitho** [reply]
in the case you want to remove multiple at once for ex ?

**Reply by KnifMelti** [reply]
Was it a question, or a thought ;) - you're the coding guy...

**Reply by Romanitho** [reply]
a question. In fact, i wondered why you wanted to be able to select multiple apps :)

**Reply by KnifMelti** [reply]
Add and remove multiple.
You never know what you find in a search

**Reply by KnifMelti** [reply]
And remove in one go to start over

**Reply by Romanitho** [reply]
[ConfigWSB.zip](https://github.com/Romanitho/Winget-Install-GUI/files/8598634/ConfigWSB.zip)

Can you try this ? :)

**Reply by KnifMelti** [reply]
Aaah, extremely nice!
Apps as a service/on demand for Windows Sandbox!!!

**Reply by Romanitho** [reply]
May I suggest something ?
We could have a WiGui like the current version for all users, and an admin version one that could install the admin tools with another tab or something (or like you do on your dev version :) ).
By the way, few of the NirSoft exist on winget, but not the Advance Run or UninstallView (yet).

**Reply by KnifMelti** [reply]
Whatever you say! 👍 
It's your project totally :)

**Reply by KnifMelti** [reply]
Ah, NirSoft installs the apps to "C:\Program Files (x86)\NirSoft" and the shorcuts to the WDAGUtilityAccount start menu - we should follow his lead (even if the shortcuts should be installed for all users)!

And AdvancedRun and UninstallView being downloaded as x64 should install to "C:\Program Files\NirSoft"

**Reply by Romanitho** [reply]
I see, this is not what you've done on your version ?

**Reply by KnifMelti** [reply]
? I just installed NirSoft 2 apps (bluescreenview and another) via WiGui (winget) and noticed where they installed...
...on the road/phone now

**Reply by KnifMelti** [reply]
Ah, now I read it Property...
...no, I just download them  and extracted to program data with links on public desktop.
All hos apps are portable

**Reply by KnifMelti** [reply]
Now fixed.

**Comment by Romanitho** [comment]
v1.4.0 out :)

**Reply by KnifMelti** [reply]
A little late response by me, but ;)

**Reply by Romanitho** [reply]
Really like the information boxes you added for prerequisites ;)

**Reply by KnifMelti** [reply]
Found a problem with ACL in **Win10 WSB** under C:\Users\WDAGUtilityAccount\AppData that prevents installation of the **MSIXBundle for App Installer**
I'll take a look later!

**Reply by Romanitho** [reply]
Sorry, long work day today ^^
Yes, I had it working under Win11 WSB, but with Win10 I observed same thing. If I have time tonight i'll have a look

**Reply by KnifMelti** [reply]
It's more complicated than ACL, it's the add-appxbundle thing (DISM) that don't create any scratch directory at all in wsb 10...
Logs under Windows/logs/dism
Loglevel can be set when calling in ps

**Reply by KnifMelti** [reply]
Too much work. Drop this pull request of mine. I'll make another when WIN11 is common...

**Reply by Romanitho** [reply]
I had it working if I replace  
`Add-AppxProvisionedPackage -Online -PackagePath "$PSScriptRoot\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle" -SkipLicense`  
by  
`Add-AppxPackage "$PSScriptRoot\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle"`

**Reply by KnifMelti** [reply]
Then one can check if win10 then use that instead.

**Reply by Romanitho** [reply]
should work with Win11 WSB as well no ?

**Reply by KnifMelti** [reply]
Well, You tell me 😁 Guess it would...

**Reply by Romanitho** [reply]
I made few changes on my dev branch, based on your work :)

edit: Forget, doesn't work :D

**Comment by KnifMelti** [comment]
I haven't had the time or energy to continue, too much at work.
I'll continue later on.
By the way, I'm now implementing Winget-Install and -AutoUpdate for the IT departments computers (25) in a pilot/POC!

**Reply by KnifMelti** [reply]
The only way I've managed to get it working under both WSB 10/11 is to include:
```
if (!(Get-AppxPackage -Name 'Microsoft.VCLibs.140.00.UWPDesktop')){
 $InfoLabel.Text = "`r`n Windows Sandbox detected:`r`n Tools and Prerequisites needed...`r`n`r`n Installing Prerequisite:`r`n Microsoft.VCLibs.140.00.UWPDesktop..."
 $InfoForm.Update()
 #Hard to find, downloaded...
 Add-AppxPackage "$PSScriptRoot\Microsoft.VCLibs.140.00.UWPDesktop_14.0.30704.0_x64__8wekyb3d8bbwe.Appx"
}
/../
Add-AppxPackage "$Location\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle"
```
So, if we can find a way to download **Microsoft.VCLibs.140.00.UWPDesktop_14.0.30704.0_x64__8wekyb3d8bbwe.Appx** from the script it'll be OK.
I've only found:
https://lifesaver.codes/answer/it-is-hard-to-find-microsoft-vclibs-140-00-uwpdesktop-3097#google_vignette
https://store.rg-adguard.net/

**Reply by Romanitho** [reply]
You can download it here no ? https://docs.microsoft.com/fr-fr/troubleshoot/developer/visualstudio/cpp/libraries/c-runtime-packages-desktop-bridge

This is the one I use

**Reply by Romanitho** [reply]
https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx

**Reply by KnifMelti** [reply]
But, it isn't 140.00.UWPDesktop...

**Reply by KnifMelti** [reply]
...so, that won't help.
That one you already handle in your code.

**Reply by Romanitho** [reply]
This is what MS wants you to install in theory

**Reply by KnifMelti** [reply]
Well, that one installs UWPDesktop too!
But you can't install it via:
```
Add-AppxPackage -Path https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx
```
You'll have to download it (and skip the check 'cause an older one is already installed):
```
#Download Microsoft.VCLibs.140.00
$VCLibsURL = 'https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx'
$WebClient=New-Object System.Net.WebClient
$WebClient.DownloadFile($VCLibsURL, "$Location\Microsoft.VCLibs.x64.14.00.Desktop.appx")
Add-AppxPackage "$Location\Microsoft.VCLibs.x64.14.00.Desktop.appx"
#Remove Microsoft.VCLibs.140.00
Remove-Item -Path "$Location\Microsoft.VCLibs.x64.14.00.Desktop.appx" -Force -ErrorAction Continue | Out-Null

```
So, now I'm gonna test on Win 10 WSB too...

**Reply by Romanitho** [reply]
I have it working for Win10 and 11 wsb

**Reply by KnifMelti** [reply]
Me too.

**Reply by Romanitho** [reply]
perfect :)

**Reply by KnifMelti** [reply]
It's all in my PR

**Reply by Romanitho** [reply]
I had issues. Maybe i did something wrong. 
But in fact, it was a mistake on the wrong detection of `Get-AppxPackage -Name 'Microsoft.VCLibs.140.00.UWPDesktop'`

**Reply by KnifMelti** [reply]
And.. your is functioning in WSB without ACL editing, must have been something then with the Prereqs not installing correctly!
Great!
Then I can delete my dev :=)
Still see that WAU when downloaded/installing is 1.8.0 and it's not detecting winget as installed...

**Reply by Romanitho** [reply]
Yes I need to change something in WAU for detection. More over, WAU can be set to run in user admin context, so I need to consider that

**Reply by KnifMelti** [reply]
Running as normal user:
```
PS $Username = "$env:Username"
PS $Username
knifmelti
PS whoami
knifmelti\knifmelti
PS $ResolveWingetPath = Resolve-Path "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_*_x64__8wekyb3d8bbwe"
PS $ResolveWingetPath
PS
```
Running as admin user:
```
PS $Username = "$env:Username"
PS $Username
knifmelti
PS whoami
knifmelti\knifmelti
PS $ResolveWingetPath = Resolve-Path "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_*_x64__8wekyb3d8bbwe"
PS $ResolveWingetPath
Path
----
C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.17.10271.0_x64__8wekyb3d8bbwe
C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.18.431.0_x64__8wekyb3d8bbwe
```
Running as System:
```
PS $Username = "$env:Username"
PS $Username
WS-Desktop$
PS whoami
nt instans\system
PS $ResolveWingetPath = Resolve-Path "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_*_x64__8wekyb3d8bbwe"
PS $ResolveWingetPath
Path
----
C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.17.10271.0_x64__8wekyb3d8bbwe
C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.18.431.0_x64__8wekyb3d8bbwe
```

**Reply by Romanitho** [reply]
Yes. So system or admin is same. And WAU is not build to run as "standard" user, as it might need admin rights to update apps...

**Reply by KnifMelti** [reply]
& whoami contains nt instans\system if running as system. $Username contains "$" (that's the computer). Thats the differense.
Cant find how to do whoami PS native

**Reply by KnifMelti** [reply]
From my perspective WAU and Winget-Install should always use --scope machine 'cause they always run in System context (Winget-Install can be user or admin in normal cases if started under WiGui).  
WiGui could have a test: if System or admin then machine, else user.. ..but it downloads Winget-Install.
Test in Winget-Install then.

**Reply by KnifMelti** [reply]
...but regarding the 1.8.0 of WAU downloaded in WiGui, isn't that why winget isn't detected?
I think I'm mixing 2 issues here...

**Reply by Romanitho** [reply]
With WiGui in fact you can add Winget arguments in fact, just tested 😅
you have to search for an app, then before submet it, you can edit. than submit:  
![image](https://user-images.githubusercontent.com/96626929/167244277-ca1609d7-67ec-4ba1-b436-5c682141f62d.png)
don't forget simple quotes `'`

**Reply by KnifMelti** [reply]
Re. Get-WingetCmd:

```
#Get WinGet Location Function
function Get-WingetCmd {
    #Get WinGet Path (if admin/system context)
    $ResolveWingetPath = Resolve-Path "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_*_x64__8wekyb3d8bbwe"
    if ($ResolveWingetPath){
        #If multiple version, pick last one
        $WingetPath = $ResolveWingetPath[-1].Path
        #Get Winget Location in System context (WinGet < 1.17)
        if (Test-Path "$WingetPath\AppInstallerCLI.exe"){
            $Script:Winget = "$WingetPath\AppInstallerCLI.exe"
        }
        #Get Winget Location in System context (WinGet > 1.17)
        elseif (Test-Path "$WingetPath\winget.exe"){
            $Script:Winget = "$WingetPath\winget.exe"
        }
    }
    else {
        #Get WinGet Location in User context
        $WingetCmd = Get-Command winget.exe -ErrorAction SilentlyContinue
        if ($WingetCmd){
            $Script:Winget = $WingetCmd.Source
        }
        else {
            Write-Log "Winget not installed!" "Red"
            break
        }
    }
    Write-Log "Using following Winget Cmd: $winget"
}

```

**Reply by KnifMelti** [reply]
Then if started under admin/system: `$Script:Scope = "machine"` else `$Script:Scope = "user"`
And in Install-App: `$Command = "install --id $AppID --scope $Scope --silent --accept-package-agreements --accept-source-agreements $AppArgs"`

**Reply by Romanitho** [reply]
the thing is, not all apps have a declared scope :/ it might break things up

**Reply by Romanitho** [reply]
for instance, I reinstalled my PC and installed Samsung.Dex in machine scope, and didn't install. without scope, works.

https://docs.microsoft.com/en-us/windows/package-manager/winget/settings#scope
Regarding this doc, it works... If machine scope is declared, try to install for system. Otherwise, use default...
The thing is: how to configure this settings file under system context :D

**Reply by KnifMelti** [reply]
Then, lets test if it breaks something if an app doesn't have a declared scope

**Reply by KnifMelti** [reply]
```
#Get WinGet Path (if admin/system context)
$ResolveWingetPath = Resolve-Path "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_*_x64__8wekyb3d8bbwe"
if ($ResolveWingetPath){
    #If multiple version, pick last one
    $WingetPath = $ResolveWingetPath[-1].Path
    #Get Winget Location in System context (WinGet < 1.17)
    if (Test-Path "$WingetPath\AppInstallerCLI.exe"){
        $Script:Winget = "$WingetPath\AppInstallerCLI.exe"
        $Script:Scope = "machine"
    }
    #Get Winget Location in System context (WinGet > 1.17)
    elseif (Test-Path "$WingetPath\winget.exe"){
        $Script:Winget = "$WingetPath\winget.exe"
        $Script:Scope = "machine"
    }
```
Install-App:
```
If ($Scope){
	$Command = "install --id $AppID --scope $Scope --silent --accept-package-agreements --accept-source-agreements $AppArgs"
}
else {
	$Command = "install --id $AppID --silent --accept-package-agreements --accept-source-agreements $AppArgs"
}

```

**Reply by Romanitho** [reply]
With settings file (even in system context) t seems to work :)
I continue my test adn try to implement something

**Reply by KnifMelti** [reply]
It just works 👍

**Reply by KnifMelti** [reply]
Ah, you're talking about the actual file "settings file"...
...it isn't needed with my proposal above.
If you're gonna edit it directly it's created in the System Profile:
C:\WINDOWS\system32\config\systemprofile\AppData\Local\Microsoft\WinGet\Settings

**Reply by Romanitho** [reply]
Yes, i've seen that. I try to find a way to edit the json file. Note easy. And i'm note a big fan of replacing it, in case some users already edited it

**Reply by KnifMelti** [reply]
But, why? my proposal takes care of it, no?
I don't think many users run as system ;)

**Reply by Romanitho** [reply]
I would like to change it fos system AND user, as you can might run WiGui as Admin User.   
I have a first draft:  
```
function test {

    $ConfigFile = Get-Content -Path "C:\Windows\System32\config\systemprofile\AppData\Local\Microsoft\WinGet\Settings\settings.json" | Where-Object {$_ -notmatch "//"} | ConvertFrom-Json

    if ($ConfigFile.installBehavior.preferences.scope){
        $ConfigFile.installBehavior.preferences.scope = "Machine"
    }
    else {
        Add-Member -InputObject $ConfigFile -MemberType NoteProperty -Name 'installBehavior' -Value $(
            New-Object PSObject -Property $(@{preferences = $(
                    New-Object PSObject -Property $(@{scope = "machine"}))
            })
        ) -Force
    }

    $ConfigFile = $ConfigFile | ConvertTo-Json | Out-File "C:\Windows\System32\config\systemprofile\AppData\Local\Microsoft\WinGet\Settings\settings.json"

}

test
```

In this way, we add setting if not set, or edit if exists. If other settings are set in the JSON file, we don't touch :)

**Reply by Romanitho** [reply]
Your proposal will try to put machine scope if run as system. The thing is, some apps doesn't know how to handle scope as it is not declared in their manifest (I guess)

**Reply by KnifMelti** [reply]
Ah, you want to seperate an Admin User and System too.. ..now I understand.
Don't forget the file doesn't exist from the beginning!

**Reply by Romanitho** [reply]
It's not easy to handle. I think we will use a predefined file. No one should change their Winget settings 😅😂

**Reply by KnifMelti** [reply]
Another proposal (yes, I know that if run as System `--scope machine` is used.. ..but I don't think it matters - I think if the scope is missing in the manifest it just installs with the manifest default (or not and then it just fails: that should be tested)):
```
#Get WinGet Location Function
function Get-WingetCmd {
    #Get WinGet Location in Admin/User context
    $WingetCmd = Get-Command winget.exe -ErrorAction SilentlyContinue
    if ($WingetCmd){
        $Script:Winget = $WingetCmd.Source
    }
    else {
        #Get WinGet Path (if System context)
        $ResolveWingetPath = Resolve-Path "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_*_x64__8wekyb3d8bbwe"
        if ($ResolveWingetPath){
            #If multiple version, pick last one
            $WingetPath = $ResolveWingetPath[-1].Path
            #Get Winget Location in System context (WinGet < 1.17)
            if (Test-Path "$WingetPath\AppInstallerCLI.exe"){
                $Script:Winget = "$WingetPath\AppInstallerCLI.exe"
                $Script:Scope = "machine"
            }
            #Get Winget Location in System context (WinGet > 1.17)
            elseif (Test-Path "$WingetPath\winget.exe"){
                $Script:Winget = "$WingetPath\winget.exe"
                $Script:Scope = "machine"
            }
        }
    }
    else {
        Write-Log "Winget not installed!" "Red"
        break
    }
    Write-Log "Using following Winget Cmd: $winget"
}
```
Install-App:
```
If ($Scope){
	$Command = "install --id $AppID --scope $Scope --silent --accept-package-agreements --accept-source-agreements $AppArgs"
}
else {
	$Command = "install --id $AppID --silent --accept-package-agreements --accept-source-agreements $AppArgs"
}

```

**Reply by Romanitho** [reply]
it fails if scope is not declared :/

**Reply by Romanitho** [reply]
but this seems to work:  
```
function Add-ScopeMachine ($path) {
    if (Test-Path $path){
        $ConfigFile = Get-Content -Path $path | Where-Object {$_ -notmatch '//'} | ConvertFrom-Json
    }
    if (!$ConfigFile){
        $ConfigFile = @{}
    }
    if ($ConfigFile.installBehavior.preferences.scope){
        $ConfigFile.installBehavior.preferences.scope = "Machine"
    }
    else {
        Add-Member -InputObject $ConfigFile -MemberType NoteProperty -Name 'installBehavior' -Value $(
            New-Object PSObject -Property $(@{preferences = $(
                    New-Object PSObject -Property $(@{scope = "machine"}))
            })
        ) -Force
    }

    $ConfigFile = $ConfigFile | ConvertTo-Json -Depth 10 | Out-File $path -Encoding utf8

}

$UserPath = "$env:LOCALAPPDATA\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\settings.json"
$SystemPath = "$Env:windir\system32\config\systemprofile\AppData\Local\Microsoft\WinGet\Settings\settings.json"

Add-ScopeMachine $UserPath
Add-ScopeMachine $SystemPath
```

**Reply by KnifMelti** [reply]
Found an app (there's not many) that doesn't use scope in the manifest: **A-SIT.PDF-Over**
Run as System with:
```
C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.18.431.0_x64__8wekyb3d8bbwe>winget install --id A-SIT.PDF-Over --scope machine --silent --accept-package-agreements --accept-source-agreements
No applicable installer found; see logs for more details.
```
winget log file:
`2022-05-07 14:25:30.090 [CLI ] Installer [X64,msi,Unknown,en-US] not applicable: Installer scope does not match required scope: Unknown != Machine
`

**Reply by KnifMelti** [reply]
Scope is **Unknown** if it doesn't exist in the installer yaml

**Reply by KnifMelti** [reply]
So, we have the same thing in both our solutions :) - NOT 🤓

**Reply by Romanitho** [reply]
in your case you got an error, no ? The apps is not installed.
If we change the scope in setting files, it uses as a "preferred" way to install

**Reply by KnifMelti** [reply]
Ah, then that's definitely better.!

**Reply by KnifMelti** [reply]
How does it work if we override with custom parameters?

**Reply by Romanitho** [reply]
if we have Scope machine in settings, but you use `--scope user` I think it will install in user scope

**Reply by KnifMelti** [reply]
Confirmed!
'Microsoft.VisualStudioCode --scope user' installs in user context, otherwise in System context (for all users under %ProgramFiiles%\Microsoft VS Code)

**Comment by KnifMelti** [comment]
Tested your Winget-Install-GUI (dev) in WSB Win11...
...**fantastic**!
Still, if installing WAU:
```
Checking if Winget is installed
Not installed. Downloading WinGet...
```
It's triggered in **Winget-AutoUpdate-Install.ps1** (from Get-GithubRepository "https://github.com/Romanitho/Winget-AutoUpdate/archive/refs/heads/main.zip"):
```
function Install-WinGet{

    #Check Package Install
    Write-Host "Checking if Winget is installed" -ForegroundColor Yellow
    $TestWinGet = Get-AppxProvisionedPackage -Online | Where-Object {$_.DisplayName -eq "Microsoft.DesktopAppInstaller"}
    If([Version]$TestWinGet.Version -gt "2022.213.0.0") {

        Write-Host "WinGet is Installed" -ForegroundColor Green
    
    }
    Else{

        #Download WinGet MSIXBundle
        Write-Host "Not installed. Downloading WinGet..."
        $WinGetURL = "https://github.com/microsoft/winget-cli/releases/download/v1.3.431/Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle"
        $WebClient=New-Object System.Net.WebClient
        $WebClient.DownloadFile($WinGetURL, "$PSScriptRoot\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle")

        #Install WinGet MSIXBundle
        try{
            Write-Host "Installing MSIXBundle for App Installer..."
            Add-AppxProvisionedPackage -Online -PackagePath "$PSScriptRoot\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle" -SkipLicense
            Write-Host "Installed MSIXBundle for App Installer" -ForegroundColor Green
        }
        catch{
            Write-Host "Failed to intall MSIXBundle for App Installer..." -ForegroundColor Red
        }
    
        #Remove WinGet MSIXBundle
        Remove-Item -Path "$PSScriptRoot\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle" -Force -ErrorAction Continue

    }

}

```

**Reply by Romanitho** [reply]
Yes, I need to add a check to avoid install when running from WiGui :)

I still need to review your PR 😅

**Reply by KnifMelti** [reply]
What PR?

**Reply by Romanitho** [reply]
on WAU

**Reply by KnifMelti** [reply]
Ah, [this](https://github.com/Romanitho/Winget-AutoUpdate/pull/74)

**Reply by KnifMelti** [reply]
Too many, too few...

**Reply by KnifMelti** [reply]
I'm sorry if I'm pushing you with my inputs/PR:s; take it easy in your own pace!
The **POC** I've implemented is working as it should!
**Thanks**!

**Reply by KnifMelti** [reply]
...regarding that; why not have the **Search an app** textbox be prechosen (typing could commence directly)?

**Reply by Romanitho** [reply]
could be nice. I have to search for how to do it :)

**Reply by KnifMelti** [reply]
$form.Add_Shown({$textBox.Select()})

**Reply by KnifMelti** [reply]
Select

**Reply by Romanitho** [reply]
done on my dev branch.

Now, review everything and implement :)

**Reply by Romanitho** [reply]
@KnifMelti https://github.com/Romanitho/Winget-Install-GUI/blob/dev/Sources/Winget-Install-GUI-AdminTools.ps1 new version for admins with possibilities to download tools to C:\Tools\

**Reply by KnifMelti** [reply]
### Really, really nice!
But it's more work for you when you're separating the **Admin Tools** from the normal (one more file to update)!
Why not in the normal one instead (those who are not admins and don't have installation rights are still going to have problems using **WiGui** for more than - i.e. **WAU** - downloading/installing programs in user context)?

My small proposals:
- Tell where the tools are being downloaded to (or open the directory)
- Check if tools are already downloaded/installed
- Uncheck the **Admin Tools** checkboxes after install a la **Configure WAU**
- Rename **CMTrace** to **ConfigMgr 2012 Toolkit**?
- Add **Configuration Manager Support Center** to **Admin Tools** (if you can find it as standalone...)

**Reply by Romanitho** [reply]
I agree, we could have only one file and run it with `-admin` switch for ex, to see the admin tab. But I wanted to separate it in another branch in fact. I clearly wanted kind of different project. In this way, most of users get the standard WiGui, and sccm or intune admin like us, get the admin version one for troubleshooting and stuff.

Regarding your proposal, I can try to implement, but my priority might be to update WAU and get everything ready before getting new wigui out 😅

Edit: regarding ConfMgr toolkit, I called it CMTrace, cause for my WAU log, this is the tool I use 😂 but i will rename it

**Reply by KnifMelti** [reply]
You're absolutely right, WAU and Winget-Install are prio 1

**Reply by Romanitho** [reply]
Made few changes on my dev wigui :)

**Reply by KnifMelti** [reply]
![16520107535706797607593129481294](https://user-images.githubusercontent.com/102996177/167294894-21c931c3-c5f1-4f9f-96a7-91283ed956d7.jpg)

**Reply by Romanitho** [reply]
haha, you're right, it's sunday 🤣

Enjoy !

**Reply by Romanitho** [reply]
Ok, enough for today. WiGui is ready.
Run `WiGui.ps1` to run normal version. Run `WiGui.ps1 -admin` and you get Admin Tools tab appear :)
Works with exe as well

**Reply by Romanitho** [reply]
What my Win11 WSB file looks like now. URL to update soon maybe :)
[WiGui Romanitho.zip](https://github.com/Romanitho/Winget-Install-GUI/files/8647221/WiGui.Romanitho.zip)

**Reply by KnifMelti** [reply]
Nice with info boxes and shortcut to CMTrace too!

**Reply by Romanitho** [reply]
I try to do things nice looking as well.

The thing is, with sandbox I cannot run the `-admin` switch in the powershell command. So I converted it to put the ps1 on the desktop. Like this we can run it with `-admin` arg

**Comment by Romanitho** [comment]
I think it's time to release the new version :p

**Reply by KnifMelti** [reply]
And, the **Winget-Install** with scope too?

**Reply by Romanitho** [reply]
yes, first
